File: 0155-min-stack/0155-min-stack.py
class MinStack(object):
    def __init__(self):
        self.stack = []
        self.min_val = []

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """
        self.stack.append(val)
        if not self.min_val or val <= self.min_val[-1]:
            self.min_val.append(val)

    def pop(self):
        """
        :rtype: None
        """
        if self.stack[-1] == self.min_val[-1]:
            self.min_val = self.min_val[:-1]
        self.stack = self.stack[:-1]

    def top(self):
        """
        :rtype: int
        """
        return self.stack[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.min_val[-1]
    
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/0155-min-stack.ts
class MinStack {
    arr: number[] = []
    minArr: number[] = []

    constructor() {
        
    }

    push(val: number): void {
        this.arr.push(val);
        if (!this.minArr.length || this.minArr[this.minArr.length-1] >= val) this.minArr.push(val);
    }

    pop(): void {
        const val = this.arr.pop();
        if (val == this.minArr[this.minArr.length-1]) this.minArr.pop();
    }

    top(): number {
        return this.arr[this.arr.length-1];   
    }

    getMin(): number {
        return this.minArr[this.minArr.length-1];
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */